mock: isolate objx-dependent code in their own source files

dolmen opened the pull request:
## Summary

In order to ease maintenance of downstream forks of Testify that would remove dependency on github.com/stretchr/objx we move all uses of that module in isolated source files. A fork without that dependency could just remove those files.
See https://github.com/stretchr/testify/issues/1752#issuecomment-2921545929

The use of objx is quite contained: it is only used in Mock.TestData(). Note that we can't just remove that method or change the return value because that would be a breaking change.

## Changes

* Move `Mock.TestData()` in a new file `mock/mock_objx.go`
* Move tests of TestData in a new file `mock/mock_objx_test.go`

## Motivation

See https://github.com/stretchr/testify/issues/1752#issuecomment-2921545929

With this change it will also be possible to add a `//go:build !testify_mock_no_objx` to `mock/mock_objx.go` and `mock/mock_objx_test.go` to allow downstream users to use testify/mock without the objx dependency.

## Related issues

* #1752

dolmen committed:
mock: isolate objx-dependent code in their own source files

In order to ease maintenance of downstream forks of Testify that would
remove dependency on github.com/stretchr/objx we move all uses of that
module in isolated source files. A fork without that dependency could
just remove those files.
See https://github.com/stretchr/testify/issues/1752#issuecomment-2921545929

The use of objx is quite contained: it is only used in Mock.TestData().
Note that we can't just remove that method or change the return value because
that would be a breaking change.

brackendawson reviewed (commented):
In my opinion I don't think this is a worthwhile change, removing my review without blocking.

This change also does the opposite of what it intends: If a fork of testify has removed objx then this change will be a merge conflict when they pull from upstream. If we touch nothing in the objx referenced code then it will not be a merge conflict for them.

brackendawson reviewed (changes requested):
This PR achieves the opposite of its intention. The stated aim is to make it easier for forks of testify to remove the objx dependency by moving code containing references to objx to new files.

Every time we touch the code referencing objx we cause a merge conflict for such forks and that includes this change. We should not make any change to make it easier for forks unless we also have to make another change to lines which reference objx.

> With this change it will be possible to add a //go:build !testify_mock_no_objx to mock/mock_objx.go and mock/mock_objx_test.go.

This will still not allow such a fork to remove objx from go.mod.

Please reference a request from the maintainer of a fork for this functionality.

ccoVeille reviewed (changes requested):
I don't think it's worth trying this

dolmen commented:
@brackendawson You wrote

> Please reference a request from the maintainer of a fork for this functionality.

Are you aware of a such fork already existing?

brackendawson commented:
I'm not aware of any. I think this feature should not be done unless it delivers good value to a customer who exists.

dolmen committed:
mock: add build tag 'testify_no_objx' to opt-out of objx

Add build tag testify_no_objx to allow to exclude method Mock.TestData()
for avoiding dependency on github.com/stretchr/objx.

dolmen committed:
CI: also run tests with build tag 'testify_no_objx'

fredbi commented:
I am discovering this piece of work.

From recent experiments at managing dependencies, I can tell for a fact that using build tags to guard dependencies does not achieve anything: consumers of the lib still get the dependencies in their go.mod.

the only viable way i found to make dependencies optional (like stuff used only in the specific use-casz of mocks) is to make an independent module.

this is doable but managing a mono repo with multiple modules requires a bit of extra attention.

i could transform the idea in that direction if you like it

ccoVeille commented:
I like your ideas for YAMLEq.

Here objx is very limited I think. It's about having a struct if I remembered well.

Can't we simply use a replace directive in our project, to force to use a local lib?

Would it work? Coukd it help to solve the dependency issue?

Yes, I agree with you the build tags things looks like an idea that brought nothing except a dead end

fredbi commented:
@ccoVeille since you know the place, take a look at github.com/go-openapi/swag. The replace directives exist only for local development and to be able to cut releases (since at this very moment you put tags in the go.mod’s that don’t exist yet)

fredbi commented:
Also notice that go.work (go1.18+) is by no means essential. It just provides a slightly improved developer experience (as it was designed for). Replace directives are essential for the multi-module with inter-dependent modules, but not resolved by the consumers of your modules

fredbi commented:
The main issue in testify is to support optional dependency without bringing a breaking change. And that part is hard.

I don’t it is reasonable to expect absolutely no breaking change (apart by reimplementing these deps with the std library only). There the testify linter could help, with an autofix to add the desired import wherever we call something that needs an extra import. Could be YAMLEq, but also the stuff mentioned above related to the mock.

on a given code base, only a tiny fraction would actually use YAMLEq so i’d expect a limited impact

ccoVeille commented:
> No replace directives only act locally not for the consumers of your package

Thanks for the confirmation. I assumed it was something like that.

> @ccoVeille since you know the place, take a look at github.com/go-openapi/swag. The replace directives exist only for local development and to be able to cut releases (since at this very moment you put tags in the go.mod’s that don’t exist yet)

👍

ccoVeille commented:
> The main issue in testify is to support optional dependency without bringing a breaking change. And that part is hard.
> 
> I don’t it is reasonable to expect absolutely no breaking change (apart by reimplementing these deps with the std library only). 

Yes, it's an option I thought about.

Import the minimal code needed to check a YAML is equivalent to another, and remove the dependency. Here it's possible because we are not exporting yaml type out of testify.

objx is different as it exposes a type from the dependency.

YAMLEq would require to copy a lot of thing, but it could be a better option as the scope is very limited.

> There the testify linter could help, with an autofix to add the desired import wherever we call something that needs an extra import. Could be YAMLEq, but also the stuff mentioned above related to the mock.
> 
> on a given code base, only a tiny fraction would actually use YAMLEq so i’d expect a limited impact 

This would be a breaking change

fredbi commented:
An intermediate position could be to indulge a into “blessed” dependencies only: objx is part of the stretchr org (or could be a module inside testify) and that’s fine by me. When consumers of my packages sees in its go.mod all the imports “stretchr/testify/xxx”  they instantly get that this is test dependencies and moves on.

my auto-merging bots can recognize those as blessed just as well.

but the yaml stuff, the spew lib and the go-difflib we should IMHO isolate somehow

fredbi commented:
Here is yet another possible strategy (i am planning to apply that one on another lib, to wit girhub.com/go-openapi/runtime, another large-bandwidth stuff we have).

There i want to cut the dependency to open telemetry. But i can't do that without a breaking change.

so what i am planning is that i’ll freeze the dependency, publish the deprecation and encourage users to use the newly published modules which is maintained with regular updates.

this is a bit similar to your idea of copying go-yaml once to ensure it works, but would relieve the maintainers from the burden of further maintaining that copy.

dolmen commented:
@brackendawson
Your review is blocking the merge. And as proposed in #1852 I've based #1870 on this one.
Could you unblock the merge so we could then merge it, and then rebase #1870?

dolmen commented:
@brackendawson With #1852 I want to drop the dependency on `objx` and move on. By having this (#1757) applied first that gives a clean fork point for downstream users (if they exist...) that still rely on `objx` features of `Mock.TestData()`.

brackendawson commented:
My review is blocking because I think this is a change which we should not make. The same feature for the yaml dependency caused forward comparability issues. It also does not deliver tangible benefit to anyone because it does not remove the dependency from the go.mod file, it only produces a different philosophical stance that the objx dependency is un-used, which users already have.

dolmen commented:
Rebased after merge of #1823 (upgrade of objx to v0.5.3).

dolmen commented:
Rebased.

I still want this to be merged.

@brackendawson 
> My review is blocking because I think this is a change which we should not make. The same feature for the yaml dependency caused forward comparability issues.

Which comparability issues?

brackendawson commented:
> Which comparability issues?

See: https://github.com/stretchr/testify/issues/1281#issuecomment-3278613428 

In short it inappropriately exposed the internal implementaiton of the yaml assertions and prevented us from fixing bugs. I will not approve any more such build modes which pretend to remove a dependency.

Edit: I did mean compatibility issues
